### ISSUE TITLE
EPIC: Enable public key rotation used to verify plugin manifests

### ISSUE THREAD

**Issue by bergquist** [grafana/grafana#23619]
Currently, the key for verifying plugin manifests is [hardcoded](https://github.com/grafana/grafana/blob/main/pkg/plugins/manager/signature/manifest.go#L34-L55) in the grafana-server. This means that if the private key is compromised people can forge manifests without us doing anything about it. If we lose the private key we cannot sign any plugins which would also be problematic. 

To solve that problem we would like to support rotating public keys. Allowing the grafana-server to download new valid keys or revoke existing keys. We currently expose all public keys that can be used to decrypt manifests in our grafana.com API `https://grafana.com/api/plugins/ci/keys`

Design doc: https://docs.google.com/document/d/1PUkqfbr3i1ZPGQ7VHE6gSbtROuuasoGXNZPdL8b76j4/edit#

**Acceptance criteria:** 
- A private key can be replaced with a new one and plugins signed with the new key will work with Grafana versions >= “this solution was introduced in”
- A private key that was leaked can be banned and all plugins signed after a specified date (date of exposure) should not pass the signature verification
- There is a solution for air-gapped instances
   - Side loading private key or falling back to the built-in one
   - This has to be validated with our cloud partners

**How to achieve:**
- Store the public keys in the database and keep it in sync with https://grafana.com/api/plugins/ci/keys every 10 days
- Have a cronjob that updates the DB every 10 days
- Extend build process to fetch the current keys (from `https://grafana.com/api/plugins/ci/keys`) at the moment of release or regularly and add them to the distribution

**Bonus points**
- Find a way to make it work for air-gapped instances, e.g. allow loading keys from a file instead 

```[tasklist]
### POC
- [x] Take a look at https://github.com/grafana/grafana/pull/23777
- [x] Add a feature flag and conditionally load the publicKey from the API
- [x] Store / retrieve the plubicKey from the object storage
- [x] Invalidate stored publicKey (or force retrieval) after a period
- [x] Improve error if the key has been revoked
```

- To evaluate: What happens if a key gets rotated? (10 days wait)
- (skipped) Control if the publicKey is loaded from the API or the hardcoded version with configuration. Not needed for the moment, until the feature flag is removed.

```[tasklist]
### Production implementation
- [x] Deploy in dev
- [x] Deploy in staging
- [x] Deploy in canary
- [x] Deploy in prod
- [ ] Remove FF, add documentation
```

**Comment by DanCech** [comment]
This is simpler than managing license enforcement keys, because there's less incentive for an end user to try to subvert the keys that are used.  I think it's reasonable for the grafana instance to get a list of keys for plugin validation from grafana.com to use for validating plugins.

**Comment by tolzhabayev** [comment]
fyi @timlevett @taleena for partner-tracking, as there were some questions about this.

**Comment by andresmgot** [comment]
After #66439, Grafana will download the public key from https://grafana.com/api/plugins/ci/keys on every boot and will store the public key in memory (this is behind a feature flag, for the moment). It will also be possible to opt-out from this behavior and simply use the hardcoded key, as of today, to make it compatible with air-gapped scenarios (this is not implemented yet). 

In this scenario, we wonder if it's still necessary to store the public key in the database and keep it in sync. The trade offs are:

Pros:
 - It will be likely faster to retrieve the key from the database than from the API. Not sure if this is relevant though comparing to other download times.

Cons:
 - If the key is in the database, even if it's periodically updated, people could modify it.
 - The extra complexity of having a local copy and implement a cron job that updates it is not trivial to implement. We would need to resolve a dependency between the plugin loading system and the storage system apart from the logic to run the cron job.

For security and simplicity reasons we think it may be more benefitial to just keep the key in memory instead. What do you think @bergquist @DanCech @tolzhabayev?

**Comment by DanCech** [comment]
The primary reason to cache the keys locally would be to avoid adding a dependency on getting the list of keys from grafana.com at startup to perform the validation, since the way it's implemented now will result in grafana being unable to load any plugins at all if it can't reach grafana.com at startup.

If we store the list of allowed keys in persistent storage (the kv store might be a good option), then we can only update the keys out of band and not block startup.

The other thing we should add to the current implementation is a documented method for revoking a key.  Right now it would be possible to return an invalid public key value for a specified key id since the verifier would error out attempting to parse the key, but it would be good to have a documented constant (e.g. REVOKED or similar) so that we can log/return a useful error message to let the operator know that the plugin isn't loading because the signing key has been revoked.

**Comment by andresmgot** [comment]
> The primary reason to cache the keys locally would be to avoid adding a dependency on getting the list of keys from grafana.com at startup to perform the validation, since the way it's implemented now will result in grafana being unable to load any plugins at all if it can't reach grafana.com at startup.
> 
> If we store the list of allowed keys in persistent storage (the kv store might be a good option), then we can only update the keys out of band and not block startup.

Yes, the proposal for this would be to fallback to the hardcoded key if grafana.com is unreachable. We can either allow this through configuration (something like `offlineKey = true`) or make it the automatic and show a warning if the online key fails to download (similar to what the update checker does). What do you think about this @DanCech, should we discard this option? (I don't know how common is to not have access to Grafana.com at startup)

**Comment by andresmgot** [comment]
> If we store the list of allowed keys in persistent storage (the kv store might be a good option), then we can only update the keys out of band and not block startup.

FTR, I explored this possibility in #66839. Let me know what you think.
